### PR TITLE
Fix ES6 decorated class double binding.

### LINKED ISSUE #5386: Class name is missing in Function when es6 target is set
I have decorator:

```typescript
export function Document() {
   return function(objectConstructor: Function) {
        console.log(objectConstructor.name); // when targeting es5 it gives me what I want - the name of my class. When targeting es6 it does not give anything
    }
}
```

Here is decorator usage:

```typescript
@Document()
export class User {
}
```

Im using "User" class name to create a document called "User", I need information how class is called. I always used es5, but now when I switched to es6 compile target I dont have information about class name anymore. Is it a bug, or it by design? If second then what is the way to get the class name?

### PR DESCRIPTION
**Pull request by rbuckton**
Creates a class body scoped alias to the class to avoid class name double binding.

Fixes #5386.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -7192,11 +7192,33 @@ namespace ts {
                 markAliasSymbolAsReferenced(symbol);
             }
 
+            const localSymbol = getExportSymbolOfValueSymbolIfExported(symbol);
```
Comment (vladima) [inline src/compiler/checker.ts]: `getExportSymbolOfValueSymbolIfExported` returns either a value that was initially passed or a merged symbol for an associated exported symbol so to me `localSymbol` is a bit misleading. In other places we call it similar values `localOrExportedSymbol`

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -7192,11 +7192,33 @@ namespace ts {
                 markAliasSymbolAsReferenced(symbol);
             }
 
+            const localSymbol = getExportSymbolOfValueSymbolIfExported(symbol);
+
+            // Due to the emit for class decorators, any reference to the class from inside of the class body
+            // must instead be rewritten to point to a temporary variable to avoid issues with the double-bind
+            // behavior of class names in ES6.
+            if (languageVersion === ScriptTarget.ES6
+                && localSymbol
```
Comment (vladima) [inline src/compiler/checker.ts]: what are cases when `localSymbol` might be `undefined`?

File: src/compiler/emitter.ts [hunk 3]
```diff
@@ -1529,13 +1532,26 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, ge
...
                             write(getGeneratedNameForNode(declaration.name));
                             return;
                         }
                     }
+                    else if (resolver.getNodeCheckFlags(node) & NodeCheckFlags.BodyScopedClassBinding) {
+                        // Due to the emit for class decorators, any reference to the class from inside of the class body
+                        // must instead be rewritten to point to a temporary variable to avoid issues with the double-bind
+                        // behavior of class names in ES6.
+                        const declaration = resolver.getReferencedValueDeclaration(node);
+                        if (declaration && declaration.kind === SyntaxKind.ClassDeclaration) {
```
Comment (vladima) [inline src/compiler/emitter.ts]: when declaration kind might not be `ClassDeclaration`? I think that if checker has set `BodyScopedClassBinding` flag it should be enough to state that referenced declaration is class

File: src/compiler/emitter.ts [hunk 4]
```diff
@@ -5106,64 +5122,114 @@ const _super = (function (geti, seti) {
...
+                        // move the export after the declaration:
                         //
-                        //     let _default = class {
-                        //     }
-                        //     _default = __decorate([dec], _default);
-                        //     export default _default;
+                        //  TypeScript                      | Javascript
+                        //  --------------------------------|------------------------------------
+                        //  @dec                            | let default_1 = class {
+                        //  export default class {          |   static x() { return default_1.y; }
```
Comment (vladima) [inline src/compiler/emitter.ts]: `export default class C` ?

Comment (rbuckton) [inline src/compiler/emitter.ts]: No. `export default class { }` is legal JavaScript. No BindingIdentifier here.

Comment (vladima) [inline src/compiler/emitter.ts]: sure, but in your example you are using `C` in `static x() { return C.y; }`

**Comment by vladima** [comment]
:+1: with a few suggestions

File: src/compiler/checker.ts (hunk 2 above)
Comment (rbuckton) [inline src/compiler/checker.ts]: None, it seems.

File: src/compiler/checker.ts (hunk 1 above)
Comment (rbuckton) [inline src/compiler/checker.ts]: Changed.

File: src/compiler/emitter.ts (hunk 3 above)
Comment (rbuckton) [inline src/compiler/emitter.ts]: I added the `BodyScopedClassBinding` flag after adding this, and neglected to simplify the logic here.

File: src/compiler/emitter.ts (hunk 4 above)
Comment (rbuckton) [inline src/compiler/emitter.ts]: Ah, I see what you mean. I'll clean up the comment.

**Comment by rbuckton** [comment]
Merged into master and cherry-picked into release-1.8.

**Pull request merged**